### PR TITLE
runners: minor cleanups and enhacements

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
This pull request doesn't do much on its own. It's meant as an easy to review way to lay some groundwork for future patches.

- clean up how runners call super() (this has just been bothering me for a while)
- add a layer of indirection between create() and the actual runner implementation, which allows us to check that common options we pass to create() are in fact supported by the runner's declared capabilities
- promotes --erase to a common option that may or may not be supported and declared via RunnerCaps

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
Just to explain why I think it's worth reviewing and merging, this is specifically groundwork for the following future patches, which I'd like to finally get around to doing for v2.4.

- Adding a common `--board-id` option (via a `board_id` capability in `RunnerCaps`) to specify which of multiple potential attached boards to flash or debug, and which we can check for and print sensible error messages about when there's missing runner support. This is already possible for various runners,  but is spelled in a bewildering variety of ways: nrfjprog has `--snr`, pyocd has `--board-id`, esp32 has `--esp-device`, bossac has `--bossac-port`, stm32flash has `--device`, blackmagicprobe has `--gdb-serial`, mdb has `--dig-device`, I'm pretty sure canopen_program has `--node-id`, and openocd has `--serial`. Yikes.
- Adding a common `--file` option and capability to say "ignore whatever the build system is telling you to flash in runners.yaml; flash this instead". This is useful when dealing with signed images for consumption by e.g. MCUboot.
- Screaming more loudly when `--erase` is provided and adding a `--no-erase` option to turn it off.
- Letting the build system add `--erase` to the default runner options whenever it thinks that is sensible. This is something I have discussed doing with @tejlmand and @oyvindronningstad on some more recent Nordic SoCs when the final image contains UICR data, since first-time users may not understand why flashing fails in that case without a complete chip erase. (The "screaming more loudly" bit is important here so that there's less "hey what happened to my bluetooth flash partition!" type confusion which happens as a result) 
- Adding `--erase` support to openocd (if possible). This is something that users of some more recent nRF SoCs would benefit from, as openocd is currently the only thread-aware debug runner for Zephyr on those SoCs, and we'd like to make --erase added by default on all the runners whenever the build system thinks it's necessary on any of them.

**Review by oyvindronningstad — APPROVED** [review]

File: scripts/west_commands/runners/canopen_program.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def do_add_parser(cls, parser):
         parser.set_defaults(confirm=True)
 
     @classmethod
-    def create(cls, cfg, args):
+    def do_create(cls, cfg, args):
```
Comment (tejlmand) [inline scripts/west_commands/runners/canopen_program.py:90]: just curious, is there some Python best practices that favors `do_create` over  `create` ?

Comment (tejlmand) [reply scripts/west_commands/runners/canopen_program.py:90]: Forget my comment, I later discovered you also have `create` and also noticed:
```
a do_create(), a new ZephyrBinaryRunner abstract class method,
and make create() itself concrete.
```

**Review by tejlmand — APPROVED** [review]
Changes looks good, ~but I don't find this in the code:~
~> Adding a common --file option and capability to say "ignore whatever the build system is telling you to flash in runners.yaml; flash this instead". This is useful when dealing with signed images for consumption by e.g. MCUboot.~

~is that in the making ?~

Never mind, must be better at reading everything in your comments.

**Review by carlescufi — APPROVED** [review]

**Pull request merged**